src/dali/plugin/pair_converter/csv_reader/kraken_csv_pricing.py

```diff
@@ -0,0 +1,205 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Kraken CSV format:
+
+import logging
+from csv import reader
+from datetime import datetime, timedelta, timezone
+from io import BytesIO
+from itertools import repeat
+from json import JSONDecodeError
+from multiprocessing.pool import ThreadPool
+from typing import Any, Dict, Iterable, List
+from zipfile import ZipFile
+
+import requests
+from requests.models import Response
+from requests.sessions import Session
+from rp2.logger import create_logger
+from rp2.rp2_decimal import RP2Decimal
+
+from dali.historical_bar import HistoricalBar
+
+# Google Drive parameters
+_ACCESS_NOT_CONFIGURED: str = "accessNotConfigured"
+_ERROR: str = "error"
+_ERRORS: str = "errors"
+_FILES: str = "files"
+_GOOGLE_API_KEY: str = "AIzaSyBPZbQdzwVAYQox79GJ8yBkKQQD9ligOf8"
+_GOOGLE_APIS_URL: str = "https://www.googleapis.com/drive/v3/files"
+_ID: str = "id"
+_KRAKEN_FOLDER_ID: str = "1aoA6SKgPbS_p3pYStXUXFvmjqShJ2jv9"
+_MESSAGE: str = "message"
+_REASON: str = "reason"
+
+# Google Drive URL Params
+_ALT: str = "alt"
+_API_KEY: str = "key"
+_CONFIRM: str = "confirm"
+_MEDIA: str = "media"
+_QUERY: str = "q"
+
+_MS_IN_SECOND: int = 1000
+
+
+class KrakenCsvPricing:
+
+    __KRAKEN_OHLCVT: str = "Kraken.com_CSVOHLCVT"
+
+    __TIMEOUT: int = 30
+    __THREAD_COUNT: int = 6
+
+    __TIMESTAMP_INDEX: int = 0
+    __OPEN: int = 1
+    __HIGH: int = 2
+    __LOW: int = 3
+    __CLOSE: int = 4
+    __VOLUME: int = 5
+    __TRADES: int = 6
+
+    __DELIMITER: str = ","
+
+    def __init__(
+        self,
+        google_api_key: str,
+    ) -> None:
+
+        self.__google_api_key: str = google_api_key
+        self.__logger: logging.Logger = create_logger(self.__KRAKEN_OHLCVT)
+        self.__session: Session = requests.Session()
+
+    def get_historical_bars_for_pair(self, base_asset: str, quote_asset: str) -> Dict[datetime, HistoricalBar]:
+        bars: Dict[datetime, HistoricalBar] = {}
+        base_file: str = f"{base_asset}_OHLCVT.zip"
+
+        self.__logger.debug(base_file)
+
+        with ZipFile(BytesIO(self._google_file_to_bytes(base_file))) as zipped_ohlcvt:
+            self.__logger.debug(zipped_ohlcvt.namelist())
+            all_timespans_for_pair: List[str] = [x for x in zipped_ohlcvt.namelist() if x.startswith(f"{base_asset}{quote_asset}_")]
+
+            if len(all_timespans_for_pair) == 0:
+                self.__logger.debug("Market not found in Kraken files. Skipping file read.")
+                return bars
+
+            with ThreadPool(self.__THREAD_COUNT) as pool:
```

Comment:
Not sure I understand the multithreaded logic: is it zipping multiple files?

Reply:
There are 6 files in each zip file one for each of the standard durations - 1m, 5m, 15m, 1h, 12h, 24h. The 1m duration has around 1 million entries. The 24h has around 1920. This multithreading unzips these 6 files and processes them in parallel.